src/asmfmt/printers/mod.rs

```diff
@@ -0,0 +1,144 @@
+    PrinterFn(
+        move |_ctx: &Context, _state: &State, f: &mut fmt::Formatter<'_>| write!(f, "{:?}", s),
+    )
+}
+
+/// Print a value using the given Rust format string.
+///
+/// Warning: formatted values are not parsable. A custom parser might need to be implemented when
+/// using `formatted` in the printer.
+pub fn formatted(s: String) -> impl Printable {
+    PrinterFn(move |_ctx: &Context, _state: &State, f: &mut fmt::Formatter<'_>| write!(f, "{}", s))
+}
+
+/// Print a list of items separated by `sep`.
+pub fn list_with_sep<T: Printable>(items: &[T], sep: ListSeparator) -> impl Printable + '_ {
+    iter_with_sep(items.iter(), sep)
+}
+
+/// Print an iterator of items separated by `sep`.
+pub fn iter_with_sep<I>(iter: I, sep: ListSeparator) -> impl Printable
+where
+    I: Iterator + Clone,
+    I::Item: Printable,
+{
+    PrinterFn(
+        move |ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>| {
+            fmt_iter(iter.clone(), ctx, state, sep, f)
+        },
+    )
+}
+
+/// Print `p` enclosed by `left` and `right`.
+pub fn enclosed<P: Printable>(left: &'static str, right: &'static str, p: P) -> impl Printable {
+    PrinterFn(
+        move |ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>| {
+            write!(f, "{}", left)?;
+            p.fmt(ctx, state, f)?;
+            write!(f, "{}", right)
+        },
+    )
+}
+
+/// Print a function type with inputs and results like `<(i32, i32) -> (i64)>`
+pub fn functional_type<'a>(
+    inputs: impl Printable + 'a,
+    results: impl Printable + 'a,
+) -> impl Printable + 'a {
+    PrinterFn(
+        move |ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>| {
+            write!(
+                f,
+                "<({}) -> ({})>",
+                inputs.print(ctx, state),
+                results.print(ctx, state)
+            )
+        },
+    )
+}
+
+/// Create a printer for IR entities that have a type.
+pub fn typed(ty: impl TypedPrinter) -> impl Printable {
+    PrinterFn(move |ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>| ty.fmt(ctx, state, f))
+}
+
+/// Used to print the type of an IR objects that are typed.
+pub trait TypedPrinter {
+    fn fmt(&self, ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>) -> fmt::Result;
+}
+
+impl TypedPrinter for dyn Type {
+    fn fmt(&self, ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        Printable::fmt(self, ctx, state, f)
+    }
+}
+
+impl<T: TypedPrinter> TypedPrinter for &T {
+    fn fmt(&self, ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (*self).fmt(ctx, state, f)
+    }
+}
+
+impl<T: Typed + ArenaObj> TypedPrinter for Ptr<T> {
+    fn fmt(&self, ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let t = self.deref(ctx).get_type(ctx);
+        Printable::fmt(&t, ctx, state, f)
+    }
+}
+
+impl<T: Typed> TypedPrinter for &Vec<T> {
```

Comment:
Updated the code by introducing an `IntoTypedPrinter` that we can implement for `IntoIterator`.

Normally I prefer not to implement traits over other traits. In that case you normally limit yourself to a single trait only and you might have a hard time to implement other spezialiations in stable rust :/

Reply:
I meant something more like this
```
impl<I> TypedPrinter for I
where
    I: Iterator + Clone,
    I::Item: Printable + Typed,
{
    fn fmt(&self, ctx: &Context, state: &State, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let sep = ListSeparator::Char(',');
        let elems = self.clone().map(|ty| ty.get_type(ctx));
        iter_with_sep(elems, sep).fmt(ctx, state, f)
    }
}
```

with the following change in where it's used:
```
functional_type(typed(self.inputs.iter()), typed(self.results.iter())).fmt(ctx, state, f)
```

This all raises a more fundamental concern to me. The function `typed` really isn't doing anything. It's just passing on a list of types to `functional_type`. More importantly, Why do we need `TypedPrinter` at all? Since we have `Typed`, and `Type` itself requires `Printable`, we should always be able to print the type of a `Typed` value. 

So we should have something like 
```
pub fn functional_type<'a>(
    inputs: &'a [Ptr<TypeObj>],
    results: &'a [Ptr<TypeObj>],
) -> impl Printable + 'a {
   ...
}
```
and called as
```
functional_type(&self.inputs, &self.results).fmt(ctx, state, f)
```

I'm not able to understand where this gets used, so maybe I'm missing something. Can you point me to it?

Reply:
The `typed` function is used by the `typed` directive ([here](https://github.com/vaivaswatha/pliron/pull/20/files#diff-5e1f4a5207f75512be872e83d887b2c6cdf21040e8f6c402179995a275c7cba8R239)).

The `typed` directive casts a value that looks like a type (or a list of types) into a `Printer`.

In case of a `Type` we want to print the value itself. In case of a function we want to print the full function type. And in case of a slice we want to use the iterator in order to print the comma separated list of types. But the macro code can not tell for sure what exactly we want to print and how, this is why we need the `trait`, to direct the cast at compile time.

For example if I have something like this:

```
struct IDFunc {
  arg: Ptr<TypeObj>,
  result: Ptr<TypeObj>,
}
```

then I want to be able to use the directive `functional_type(typed($arg), typed($result))` like so.

An example using Vec:

```
struct SumAll {
  args: Vec<Ptr<TypeObj>>,
  result: Ptr<TypeObj>,
}
```

and the directive ``functional_type(typed($args), typed($result))`.

If we are not sure yet about how to properly implement the `typed` directive, let's remove it from this PR for now. Most directives are currently inspired by MLIR, but that doesn't mean that we need all of them on day 1.

Let's rather focus on getting the "infrastructure" (proc macros) in. Once we start using the new functionality we will see which pattern emerge and where problems are.

WDYT?

Reply:
Yes, let's not have that trait now. I'm not fully convinced (or not understanding it's need right). The main infrastructure is definitely the priority.